src/Lykke.AlgoStore.MatchingEngineAdapter.Services/Listening/ConsumingWorker.cs

```diff
@@ -0,0 +1,133 @@
+
+            _messageHandlers.Add(typeof(PingRequest), PingHandler);
+            _messageHandlers.Add(typeof(MarketOrderRequest), MarketOrderRequestHandler);
+
+            _thread = new Thread(DoWork) { Priority = ThreadPriority.Highest };
+            _thread.Start(_cts.Token);
+        }
+
+        /// <summary>
+        /// Disposes the <see cref="ConsumingWorker"/>
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (_isDisposed) return;
+
+            if (isDisposing)
+            {
+                _cts.Cancel();
+                _thread.Interrupt();
+                _thread.Join();
+            }
+
+            _isDisposed = true;
+        }
+
+        private void DoWork(object cancellationTokenObj)
+        {
+            var cancellationToken = (CancellationToken)cancellationTokenObj;
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    var request = _requestQueue.Dequeue(cancellationToken);
+                    Interlocked.Exchange(ref _lastMessageTime, DateTime.Now.Ticks);
+
+                    var messageType = request.Message.GetType();
+
+                    if (_messageHandlers.ContainsKey(messageType))
+                        _messageHandlers[messageType](request);
+                }
+                catch (OperationCanceledException exception)
+                {
+                    _log.WriteErrorAsync(nameof(ConsumingWorker), nameof(DoWork), null, exception).Wait();
+                    return;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Handles a <see cref="PingRequest"/> by replying with <see cref="MeaResponseType.Pong"/> containing
+        /// the same message
+        /// </summary>
+        /// <param name="request">The <see cref="MessageInfo"/> containing the message</param>
+        private void PingHandler(IMessageInfo request)
+        {
+            var msg = (PingRequest)request.Message;
+
+            request.Reply(MeaResponseType.Pong, msg);
+        }
+
+
+        /// <summary>
+        /// Handles a <see cref="MarketOrderRequest"/> by replying with <see cref="ResponseModel{T}"/> containing
+        /// the response message />
+        /// </summary>
+        /// <param name="request">The <see cref="MessageInfo"/> containing the message</param>
+        private void MarketOrderRequestHandler(IMessageInfo request)
+        {
+            var msg = (MarketOrderRequest)request.Message;
+
+            var result = _matchingEngineAdapter.HandleMarketOrderAsync(msg.ClientId, msg.AssetPairId, msg.OrderAction,
+                msg.Volume, msg.IsStraight, msg.InstanceId);
+
+            request.Reply(MeaResponseType.MarketOrderResponse, result.Result);
```

Comment:
Thread lock. Consider using non-blocking approach. Use async/await everywhere

Reply:
What is correct url of sonnarqube, I can't open it
![image](https://user-images.githubusercontent.com/31039385/39911613-b564fe12-5504-11e8-993a-5ac573af04f9.png)

Reply:
![image](https://user-images.githubusercontent.com/14345655/39911921-4b6aa928-54fd-11e8-85bd-fe0834b1fe03.png)
I get certificate error when I try to use provided sonnarcube link